File: 13. RomanToInteger/RomanToInteger(1).js
// Runtime: 264 ms, faster than 14.92% of JavaScript online submissions for Roman to Integer.
// Memory Usage: 42.1 MB, less than 63.92% of JavaScript online submissions for Roman to Integer.
/**
 * @param {string} s
 * @return {number}
 */
var romanToInt = function(s) {
    var decimal = 0;
    for (var i=s.length-1; i>=0; i--){
        if (s[i] == 'I'){
            decimal++;
            continue;
        }

        if (s[i] == 'V'){
            decimal += 5;
            if (i > 0 && s[i-1] == 'I'){
                decimal--;
                i--;
            }
            continue;
        }
        if (s[i] == 'X'){
            decimal += 10;
            if (i > 0 && s[i-1] == 'I'){
                decimal--;
                i--;
            }
            continue;
        }

        if (s[i] == 'L'){
            decimal += 50;
            if (i > 0 && s[i-1] == 'X'){
                decimal -= 10;
                i--;
            }
            continue;
        }
        if (s[i] == 'C'){
            decimal += 100;
            if (i > 0 && s[i-1] == 'X'){
                decimal -= 10;
                i--;
            }
            continue;
        }

        if (s[i] == 'D'){
            decimal += 500;
            if (i > 0 && s[i-1] == 'C'){
                decimal -= 100;
                i--;
            }
            continue;
        }
        if (s[i] == 'M'){
            decimal += 1000;
            if (i > 0 && s[i-1] == 'C'){
                decimal -= 100;
                i--;
            }
            continue;
        }
    }
    return decimal;
};

File: 13. RomanToInteger/RomanToInteger(1).py
# Runtime: 72 ms, faster than 19.25% of Python3 online submissions for Roman to Integer.
# Memory Usage: 13.9 MB, less than 31.53% of Python3 online submissions for Roman to Integer.
class Solution:
    def romanToInt(self, s: str) -> int:
        map_dict = {'I': 1, 
                    'V': 5, 
                    'X': 10, 
                    'L': 50, 
                    'C': 100, 
                    'D': 500, 
                    'M': 1000}

        decimal = 0
        i = len(s)-1
        while i >= 0:
            decimal = decimal + map_dict[s[i]]
            if (i > 0 and map_dict[s[i-1]] < map_dict[s[i]]):
                decimal = decimal - map_dict[s[i-1]]
                i = i - 1
            i = i - 1
        return decimal